pkg/network/cache/multi.go

```diff
@@ -231,25 +232,35 @@ func (x *multiClient) Close() error {
 }
 
 func (x *multiClient) RawForAddress(addr network.Address, f func(client *rawclient.Client) error) error {
-	return x.client(addr).ExecRaw(f)
+	c, err := x.client(addr)
+	if err != nil {
+		return err
+	}
+	return c.ExecRaw(f)
 }
 
-func (x *multiClient) client(addr network.Address) clientcore.Client {
+func (x *multiClient) client(addr network.Address) (clientcore.Client, error) {
 	strAddr := addr.String()
 
 	x.mtx.RLock()
 	c, cached := x.clients[strAddr]
 	x.mtx.RUnlock()
 
-	if !cached {
-		x.mtx.Lock()
+	if cached {
+		return c, nil
+	}
+
+	x.mtx.Lock()
+	defer x.mtx.Unlock()
 
-		c, cached = x.clients[strAddr]
-		if !cached {
-			c = x.createForAddress(addr)
+	c, cached = x.clients[strAddr]
+	if !cached {
+		c, err := x.createForAddress(addr)
+		if err != nil {
+			delete(x.clients, strAddr)
```

Comment:
Why do we need this? As I can see if `createForAddress` fails, there is no map entry. Also seems like map write below is redundant, too. May be you wanted to avoid caching clients in `createForAddress`?